rafttest: TestRestart is flaky

`TestRestart` failed (when run many times) at commit ed26e90 with the following log:

```
=== RUN   TestRestart
raft2024/03/08 13:35:12 INFO: 1 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 1 became follower at term 0
raft2024/03/08 13:35:12 INFO: newRaft 1 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/03/08 13:35:12 INFO: 1 became follower at term 1
raft2024/03/08 13:35:12 INFO: 1 switched to configuration voters=(1)
raft2024/03/08 13:35:12 INFO: 1 switched to configuration voters=(1 2)
raft2024/03/08 13:35:12 INFO: 1 switched to configuration voters=(1 2 3)
raft2024/03/08 13:35:12 INFO: 1 switched to configuration voters=(1 2 3 4)
raft2024/03/08 13:35:12 INFO: 1 switched to configuration voters=(1 2 3 4 5)
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 2 became follower at term 0
raft2024/03/08 13:35:12 INFO: newRaft 2 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/03/08 13:35:12 INFO: 2 became follower at term 1
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=(1)
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=(1 2)
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=(1 2 3)
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=(1 2 3 4)
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=(1 2 3 4 5)
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 3 became follower at term 0
raft2024/03/08 13:35:12 INFO: newRaft 3 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/03/08 13:35:12 INFO: 3 became follower at term 1
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=(1)
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=(1 2)
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=(1 2 3)
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=(1 2 3 4)
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=(1 2 3 4 5)
raft2024/03/08 13:35:12 INFO: 4 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 4 became follower at term 0
raft2024/03/08 13:35:12 INFO: newRaft 4 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/03/08 13:35:12 INFO: 4 became follower at term 1
raft2024/03/08 13:35:12 INFO: 4 switched to configuration voters=(1)
raft2024/03/08 13:35:12 INFO: 4 switched to configuration voters=(1 2)
raft2024/03/08 13:35:12 INFO: 4 switched to configuration voters=(1 2 3)
raft2024/03/08 13:35:12 INFO: 4 switched to configuration voters=(1 2 3 4)
raft2024/03/08 13:35:12 INFO: 4 switched to configuration voters=(1 2 3 4 5)
raft2024/03/08 13:35:12 INFO: 5 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 5 became follower at term 0
raft2024/03/08 13:35:12 INFO: newRaft 5 [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
raft2024/03/08 13:35:12 INFO: 5 became follower at term 1
raft2024/03/08 13:35:12 INFO: 5 switched to configuration voters=(1)
raft2024/03/08 13:35:12 INFO: 5 switched to configuration voters=(1 2)
raft2024/03/08 13:35:12 INFO: 5 switched to configuration voters=(1 2 3)
raft2024/03/08 13:35:12 INFO: 5 switched to configuration voters=(1 2 3 4)
raft2024/03/08 13:35:12 INFO: 5 switched to configuration voters=(1 2 3 4 5)
raft2024/03/08 13:35:12 INFO: 3 is starting a new election at term 1
raft2024/03/08 13:35:12 INFO: 3 became candidate at term 2
raft2024/03/08 13:35:12 INFO: 3 [logterm: 1, index: 5] sent MsgVote request to 1 at term 2
raft2024/03/08 13:35:12 INFO: 3 [logterm: 1, index: 5] sent MsgVote request to 2 at term 2
raft2024/03/08 13:35:12 INFO: 3 [logterm: 1, index: 5] sent MsgVote request to 4 at term 2
raft2024/03/08 13:35:12 INFO: 3 [logterm: 1, index: 5] sent MsgVote request to 5 at term 2
raft2024/03/08 13:35:12 INFO: 3 received MsgVoteResp from 3 at term 2
raft2024/03/08 13:35:12 INFO: 3 has received 1 MsgVoteResp votes and 0 vote rejections
raft2024/03/08 13:35:12 INFO: 5 [term: 1] received a MsgVote message with higher term from 3 [term: 2]
raft2024/03/08 13:35:12 INFO: 5 became follower at term 2
raft2024/03/08 13:35:12 INFO: 5 [logterm: 1, index: 5, vote: 0] cast MsgVote for 3 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 2 is starting a new election at term 1
raft2024/03/08 13:35:12 INFO: 2 became candidate at term 2
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5] sent MsgVote request to 1 at term 2
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5] sent MsgVote request to 3 at term 2
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5] sent MsgVote request to 4 at term 2
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5] sent MsgVote request to 5 at term 2
raft2024/03/08 13:35:12 INFO: 4 is starting a new election at term 1
raft2024/03/08 13:35:12 INFO: 4 became candidate at term 2
raft2024/03/08 13:35:12 INFO: 4 [logterm: 1, index: 5] sent MsgVote request to 1 at term 2
raft2024/03/08 13:35:12 INFO: 4 [logterm: 1, index: 5] sent MsgVote request to 2 at term 2
raft2024/03/08 13:35:12 INFO: 4 [logterm: 1, index: 5] sent MsgVote request to 3 at term 2
raft2024/03/08 13:35:12 INFO: 4 [logterm: 1, index: 5] sent MsgVote request to 5 at term 2
raft2024/03/08 13:35:12 INFO: 4 received MsgVoteResp from 4 at term 2
raft2024/03/08 13:35:12 INFO: 4 has received 1 MsgVoteResp votes and 0 vote rejections
raft2024/03/08 13:35:12 INFO: 2 received MsgVoteResp from 2 at term 2
raft2024/03/08 13:35:12 INFO: 2 has received 1 MsgVoteResp votes and 0 vote rejections
raft2024/03/08 13:35:12 INFO: 4 [logterm: 1, index: 5, vote: 4] rejected MsgVote from 3 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 3 received MsgVoteResp rejection from 4 at term 2
raft2024/03/08 13:35:12 INFO: 3 has received 1 MsgVoteResp votes and 1 vote rejections
raft2024/03/08 13:35:12 INFO: 5 [logterm: 1, index: 5, vote: 3] rejected MsgVote from 2 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 1 [term: 1] received a MsgVote message with higher term from 3 [term: 2]
raft2024/03/08 13:35:12 INFO: 1 became follower at term 2
raft2024/03/08 13:35:12 INFO: 1 [logterm: 1, index: 5, vote: 0] cast MsgVote for 3 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5, vote: 2] rejected MsgVote from 3 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 3 received MsgVoteResp from 5 at term 2
raft2024/03/08 13:35:12 INFO: 3 has received 2 MsgVoteResp votes and 1 vote rejections
raft2024/03/08 13:35:12 INFO: 1 [logterm: 1, index: 5, vote: 3] rejected MsgVote from 2 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 4 [logterm: 1, index: 5, vote: 4] rejected MsgVote from 2 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 5 [logterm: 1, index: 5, vote: 3] rejected MsgVote from 4 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 3 received MsgVoteResp from 1 at term 2
raft2024/03/08 13:35:12 INFO: 3 has received 3 MsgVoteResp votes and 1 vote rejections
raft2024/03/08 13:35:12 INFO: 3 became leader at term 2
raft2024/03/08 13:35:12 INFO: raft.node: 3 elected leader 3 at term 2
raft2024/03/08 13:35:12 INFO: 1 [logterm: 1, index: 5, vote: 3] rejected MsgVote from 4 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 3 [logterm: 2, index: 6, vote: 3] rejected MsgVote from 4 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: raft.node: 1 elected leader 3 at term 2
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5, vote: 2] rejected MsgVote from 4 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: 3 [logterm: 2, index: 6, vote: 3] rejected MsgVote from 2 [logterm: 1, index: 5] at term 2
raft2024/03/08 13:35:12 INFO: raft.node: 5 elected leader 3 at term 2
raft2024/03/08 13:35:12 INFO: 2 received MsgVoteResp rejection from 5 at term 2
raft2024/03/08 13:35:12 INFO: 2 has received 1 MsgVoteResp votes and 1 vote rejections
2024/03/08 13:35:12 raft.2: stop
raft2024/03/08 13:35:12 INFO: 4 received MsgVoteResp rejection from 5 at term 2
raft2024/03/08 13:35:12 INFO: 4 has received 1 MsgVoteResp votes and 1 vote rejections
raft2024/03/08 13:35:12 INFO: 4 became follower at term 2
raft2024/03/08 13:35:12 INFO: raft.node: 4 elected leader 3 at term 2
2024/03/08 13:35:12 raft.3: stop
raft2024/03/08 13:35:12 INFO: 3 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 3 became follower at term 2
raft2024/03/08 13:35:12 INFO: newRaft 3 [peers: [], term: 2, commit: 5, applied: 0, lastindex: 8, lastterm: 2]
raft2024/03/08 13:35:12 INFO: 2 switched to configuration voters=()
raft2024/03/08 13:35:12 INFO: 2 became follower at term 2
raft2024/03/08 13:35:12 INFO: newRaft 2 [peers: [], term: 2, commit: 5, applied: 0, lastindex: 5, lastterm: 1]
raft2024/03/08 13:35:12 INFO: raft.node: 2 elected leader 3 at term 2
raft2024/03/08 13:35:12 INFO: 1 is starting a new election at term 2
raft2024/03/08 13:35:12 INFO: 1 became candidate at term 3
raft2024/03/08 13:35:12 INFO: 1 [logterm: 2, index: 8] sent MsgVote request to 2 at term 3
raft2024/03/08 13:35:12 INFO: 1 [logterm: 2, index: 8] sent MsgVote request to 3 at term 3
raft2024/03/08 13:35:12 INFO: 1 [logterm: 2, index: 8] sent MsgVote request to 4 at term 3
raft2024/03/08 13:35:12 INFO: 1 [logterm: 2, index: 8] sent MsgVote request to 5 at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 1 lost leader 3 at term 3
raft2024/03/08 13:35:12 INFO: 1 received MsgVoteResp from 1 at term 3
raft2024/03/08 13:35:12 INFO: 1 has received 1 MsgVoteResp votes and 0 vote rejections
raft2024/03/08 13:35:12 INFO: 4 [term: 2] received a MsgVote message with higher term from 1 [term: 3]
raft2024/03/08 13:35:12 INFO: 4 became follower at term 3
raft2024/03/08 13:35:12 INFO: 4 [logterm: 2, index: 6, vote: 0] cast MsgVote for 1 [logterm: 2, index: 8] at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 4 lost leader 3 at term 3
raft2024/03/08 13:35:12 INFO: 3 [term: 2] received a MsgVote message with higher term from 1 [term: 3]
raft2024/03/08 13:35:12 INFO: 3 became follower at term 3
raft2024/03/08 13:35:12 INFO: 3 [logterm: 2, index: 8, vote: 0] cast MsgVote for 1 [logterm: 2, index: 8] at term 3
raft2024/03/08 13:35:12 INFO: 1 received MsgVoteResp from 4 at term 3
raft2024/03/08 13:35:12 INFO: 1 has received 2 MsgVoteResp votes and 0 vote rejections
raft2024/03/08 13:35:12 INFO: 2 [term: 2] received a MsgVote message with higher term from 1 [term: 3]
raft2024/03/08 13:35:12 INFO: 2 became follower at term 3
raft2024/03/08 13:35:12 INFO: 2 [logterm: 1, index: 5, vote: 0] cast MsgVote for 1 [logterm: 2, index: 8] at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 2 lost leader 3 at term 3
raft2024/03/08 13:35:12 INFO: 1 received MsgVoteResp from 3 at term 3
raft2024/03/08 13:35:12 INFO: 1 has received 3 MsgVoteResp votes and 0 vote rejections
raft2024/03/08 13:35:12 INFO: 1 became leader at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 1 elected leader 1 at term 3
raft2024/03/08 13:35:12 INFO: 5 [term: 2] received a MsgVote message with higher term from 1 [term: 3]
raft2024/03/08 13:35:12 INFO: 5 became follower at term 3
raft2024/03/08 13:35:12 INFO: 5 [logterm: 2, index: 6, vote: 0] cast MsgVote for 1 [logterm: 2, index: 8] at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 5 lost leader 3 at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 4 elected leader 1 at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 2 elected leader 1 at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 3 elected leader 1 at term 3
raft2024/03/08 13:35:12 INFO: raft.node: 5 elected leader 1 at term 3
    node_test.go:83: commits failed to converge!
--- FAIL: TestRestart (5.13s)
```

Looking at the test implementation, I suspect the problem is in the [waitLeader](https://github.com/etcd-io/raft/blob/bc285dd30aad36308c375b9a0ef26483aaeb72a3/rafttest/node_test.go#L130) function. It waits for a wrong signal, and erroneously reports some non-leader node to be the leader. The `n.Status().SoftState.Lead` that it collects from all the nodes only represents votes, but the actual elected leader can end up different.

A fix would be to wait for a node that is in `StateLeader`.

--- a/rafttest/node_test.go
+++ b/rafttest/node_test.go
@@ -129,7 +129,7 @@ func TestPause(t *testing.T) {
 
 func waitLeader(ns []*node) int {
 	var l map[uint64]struct{}
-	var lindex int
+	var lindex = -1
 
 	for {
 		l = make(map[uint64]struct{})
@@ -144,7 +144,7 @@ func waitLeader(ns []*node) int {
 			}
 		}
 
-		if len(l) == 1 {
+		if len(l) == 1 && lindex != -1 {
 			return lindex
 		}
 	}

